node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+	}
+}
+
+impl ProvisioningJob {
+	pub fn new(relay_parent: Hash, sender: mpsc::Sender<FromJob>, receiver: mpsc::Receiver<ToJob>) -> Self {
+		Self {
+			relay_parent,
+			sender,
+			receiver,
+			provisionable_data_channels: Vec::new(),
+			backed_candidates: Vec::new(),
+			signed_bitfields: Vec::new(),
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			use ProvisionerMessage::{RequestBlockAuthorshipData, RequestInherentData, ProvisionableData};
+
+			match msg {
+				ToJob::Provisioner(RequestInherentData(_, sender)) => self.send_inherent_data(sender).await?,
+				ToJob::Provisioner(RequestBlockAuthorshipData(_, sender)) => self.provisionable_data_channels.push(sender),
+				ToJob::Provisioner(ProvisionableData(data)) => {
+					for channel in self.provisionable_data_channels.iter_mut() {
+						// REVIEW: the try operator here breaks the run loop if any receiver ever unexpectedly
+						// closes their channel. Is that desired?
+						channel.send(data.clone()).await?;
+					}
+					self.note_provisionable_data(data);
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	// REVIEW: not async, but small; hopefully permissable
+	fn note_provisionable_data(&mut self, provisionable_data: ProvisionableData) {
+		use ProvisionableData::{Bitfield, BackedCandidate};
+
+		match provisionable_data {
+			Bitfield(_, signed_bitfield) => self.signed_bitfields.push(signed_bitfield),
+			BackedCandidate(backed_candidate) => self.backed_candidates.push(backed_candidate),
+			_ => {}
+		}
+	}
+
+	// The provisioner is the subsystem best suited to choosing which specific
+	// backed candidates and availability bitfields should be assembled into the
+	// block. To engage this functionality, a
+	// `ProvisionerMessage::RequestInherentData` is sent; the response is a set of
+	// non-conflicting candidates and the appropriate bitfields. Non-conflicting
+	// means that there are never two distinct parachain candidates included for
+	// the same parachain and that new parachain candidates cannot be included
+	// until the previous one either gets declared available or expired.
+	//
+	// The main complication here is going to be around handling
+	// occupied-core-assumptions. We might have candidates that are only
+	// includable when some bitfields are included. And we might have candidates
+	// that are not includable when certain bitfields are included.
+	//
+	// When we're choosing bitfields to include, the rule should be simple:
+	// maximize availability. So basically, include all bitfields. And then
+	// choose a coherent set of candidates along with that.
+	async fn send_inherent_data(&mut self, sender: oneshot::Sender<ProvisionerInherentData>) -> Result<(), Error> {
+
+		// coherency rules for backed candidates, recap:
+		//
+		// - [x] only one per parachain
+		// - [ ] candidate is available (2/3+ of validators have set the 1 bit)
+		// - [ ] not expired (how can we tell?)
```

Comment:
Say more on this topic? I wasn't aware of any kind of expiry

Reply:
https://w3f.github.io/parachain-implementers-guide/node/utility/provisioner.html#block-production

> new parachain candidates cannot be included until the previous one either gets declared available or expired.

This implies two checks:

- previous parachain candidate _did_ expire
- current parachain candidate _did not_ expire

But, as I said in the comment: I don't actually know how to tell what's expired.

Reply:
The recipe you wrote in the previous comment implies that we don't need any kind of expiry check here anyway, so I'm going to be proceeding on that assumption for now.